### ISSUE TITLE
roslaunch "find" syntax behaves strangely when an executable has the same name as a directory 

### ISSUE THREAD

**Issue by jonbinney** [ros/ros_comm#506]
In install space, if an executable script shares the same name as something (a directory or plain file or whatever) that gets installed to share, then the executable script gets found first when using `$(find)` in a roslaunch file, in ways that will surprise users.

Example: consider a package "bar" which has an executable script called "baz" and also has a plain file called "baz" that gets installed to share/bar/baz.

```
<param name="foo" value="$(find bar)"/>
```

Results in foo having the value

```
.../install/share/bar as expected
```

but

```
<param name="foo" value="$(find bar)/x/y/z/baz"/>   
```

results in foo having the value

```
.../install/lib/bar/baz
```

This is especially surprising since the "x/y/z" bit has been completely dropped. I know that because the install layout changed from rosmake to catkin, preserving the "find" syntax has been difficult, but this behavior isn't at all documented in http://wiki.ros.org/roslaunch/XML

It also seems that "x/y/z" shouldn't be dropped when looking for the executable, but maybe that is necessary to handle some other case?

**Comment by chadrockey** [comment]
@jonbinney I think the x/y/z part is dropped because you could have executables in arbitrary directories anywhere in the package structure.

Example:

from tutorial: catkin_install_python(PROGRAMS scripts/myscript
DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION})

baz example: catkin_install_python(PROGRAMS x/y/z/baz 
DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION})

This is a valid case of dropping the x y z baz because 'baz' must go into the CATKIN_PACKAGE_BIN_DESTINATION.

I can't think of a great solution to this.  My best thought would be to give an error when there are multiple matches to the rule.  I think this has to be an execution stopping error (as of j-turtle, Indigo can only give a very strongly worded warning).

Then I guess add an argument 'find -exec' or 'find -shr' or 'find -dir' and only handle their portion of the functionality?

**Comment by dirk-thomas** [comment]
It is exactly as @chadrockey described it.

That "weird" behavior was implemented in #233 to at least resolve the eminent problem.that scripts where not found.

I proposed more specific commands in #228 but it was not possible to reach a consensus on them.

@jonbinney If this answers your question can you please close this ticket?

**Comment by jonbinney** [comment]
Sure
